### PR TITLE
Throw different expection when EADDRNOTAVAIL

### PR DESCRIPTION
**Pull request by lillo42**
fixes: https://github.com/redhat-developer/kestrel-linux-transport/issues/91

Test in docker with ASP.NET Core:
 - [x] 3.0
 - [x] 2.1
 - [ ] 2.2

### CODE REVIEW

**Comment by lillo42** [comment]
The problem throw using ```AddressInUseException``` is the ASP.NET Core rethrow with IOException:

https://github.com/aspnet/AspNetCore/blob/dfba024c7888e9357ce40687e1771953074340fd/src/Servers/Kestrel/Core/src/Internal/AddressBinder.cs#L101-L113

```c#
internal static async Task BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
{
    try
    {
        await context.CreateBinding(endpoint).ConfigureAwait(false);
    }
    catch (AddressInUseException ex)
    {
        throw new IOException(CoreStrings.FormatEndpointAlreadyInUse(endpoint), ex);
    }
    context.ListenOptions.Add(endpoint);
}
```

File: src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/AddressNotAvailableException.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+using System;
+using System.Runtime.Serialization;
+
+namespace RedHat.AspNetCore.Server.Kestrel.Transport.Linux
+{
+    public class AddressNotAvailableException : Exception
```
Comment (tmds) [inline src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/AddressNotAvailableException.cs]: Use this one instead: https://github.com/aspnet/AspNetCore/blob/0da5fabdbdd462cdd9528398efb6b1ffb9b27f65/src/Servers/Connections.Abstractions/src/Exceptions/AddressInUseException.cs#L8

File: src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/PosixResult.Values.cs [hunk 2]
```diff
@@ -16,5 +16,8 @@ internal partial struct PosixResult
         public static int ENOBUFS => -Tmds.Linux.LibC.ENOBUFS;
         public static int EPIPE => -Tmds.Linux.LibC.EPIPE;
         public static int ECONNREFUSED => -Tmds.Linux.LibC.ECONNREFUSED;
+        public static int EADDRINUSE => -Tmds.Linux.LibC.EADDRINUSE;
+        
+        private const string AddressNotAvailable = "EADDRNOTAVAIL";
```
Comment (tmds) [inline src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/PosixResult.Values.cs]: This string isn't needed.

File: src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/PosixResult.cs [hunk 3]
```diff
@@ -40,20 +41,17 @@ internal string ErrorDescription()
             {
                 return string.Empty;
             }
-            else
+
+            lock (s_errnoDescriptions)
```
Comment (tmds) [inline src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/PosixResult.cs]: Can you revert the changes to `PosixResult`?

File: src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/AddressNotAvailableException.cs (hunk 1 above)
Comment (lillo42) [reply src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/AddressNotAvailableException.cs]: If I use ```AddressInUseException```

ASP.NET Core will rethrow with ```IOException```

See:
https://github.com/aspnet/AspNetCore/blob/4ef204e13b88c0734e0e94a1cc4c0ef05f40849e/src/Servers/Kestrel/Core/src/Internal/AddressBinder.cs#L101-L113

File: src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/TransportThread.TSocket.cs [hunk 4]
```diff
@@ -883,14 +883,19 @@ public unsafe PosixResult TrySetSocketOption(int level, int optname, int value)
                 return SocketInterop.SetSockOpt(Fd, level, optname, (byte*)&value, 4);
             }
 
-            public unsafe void Bind(IPEndPointStruct endpoint)
+            public void Bind(IPEndPointStruct endpoint)
```
Comment (tmds) [inline src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/TransportThread.TSocket.cs]: Is someone still calling this? If not, it can be removed.

File: src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/TransportThread.ThreadContext.cs [hunk 5]
```diff
@@ -196,7 +197,15 @@ private TSocket CreateAcceptSocket(IPEndPoint endPoint, SocketFlags flags)
                         }
                     }
 
-                    acceptSocket.Bind(endPoint);
+                    var bindResult = acceptSocket.TryBind(endPoint);
+
+                    if (bindResult == PosixResult.EADDRINUSE)
+                    {
+                        throw new AddressInUseException("EADDRINUSE");
```
Comment (tmds) [inline src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/TransportThread.ThreadContext.cs]: "EADDRINUSE" -> "Address in use."

File: src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/AddressNotAvailableException.cs (hunk 1 above)
Comment (lillo42) [reply src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/AddressNotAvailableException.cs]: Should I do this?
```suggestion
    public class AddressNotAvailableException : InvalidOperationException
```

Comment (tmds) [reply src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/AddressNotAvailableException.cs]: > If I use AddressInUseException
ASP.NET Core will rethrow as IOException
See:
https://github.com/aspnet/AspNetCore/blob/4ef204e13b88c0734e0e94a1cc4c0ef05f40849e/src/Servers/Kestrel/Core/src/Internal/AddressBinder.cs#L101-L113

Have you tried? I think that code path isn't used here, because the AspNetCore Libuv Transport is throwing AddressInUseException.

Comment (lillo42) [reply src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/AddressNotAvailableException.cs]: I try with ASP.NET Core 2.1 and if I use ```AddressInUseException```  I  continue with error, I didn't try with ASP.NET Core 3.0 because I only can reproduce it in docker and .NET Core 3.0 + Docker + Linux 5.0.0-29 has a bug :(

Comment (lillo42) [reply src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/AddressNotAvailableException.cs]: In Libuv, they throw when the error  is ```LibuvConstants.EADDRINUSE``` in ours case is ```PosixResult.EADDRINUSE``` already? Or ```PosixResult.EADDRNOTAVAIL`` minding same thing?

Comment (tmds) [reply src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/AddressNotAvailableException.cs]: Ah, yes. They are different.

File: src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/TransportThread.TSocket.cs [hunk 6]
```diff
@@ -882,15 +882,15 @@ public unsafe PosixResult TrySetSocketOption(int level, int optname, int value)
             {
                 return SocketInterop.SetSockOpt(Fd, level, optname, (byte*)&value, 4);
             }
-
-            public unsafe void Bind(IPEndPointStruct endpoint)
+
```
Comment (tmds) [inline src/RedHat.AspNetCore.Server.Kestrel.Transport.Linux/TransportThread.TSocket.cs]: nit: remove whitespace

**Comment by tmds** [comment]
CI fail is known issue: https://github.com/redhat-developer/kestrel-linux-transport/issues/90

I've triggered a new build.

**Comment by tmds** [comment]
Thanks @lillo42 for fixing this!!
